### ISSUE TITLE
Unexpected, batch size and device dependent NaN propagation in Conv1d

### ISSUE THREAD

**Issue by markmbaum** [pytorch/pytorch#157237]
### 🐛 Describe the bug

I'm encountering unexpected NaN propagation as I test causal 1d convolutions and the problem depends on the batch size and device.

When some sequences in a batch have NaN portions at their ends, the pad -> conv1d operation produces extra NaNs at the boundary of NaN and non-NaN elements. This is a complete, minimal example that produces the problem:

```python
import torch

torch.manual_seed(1)

device = torch.device("cpu")

# convolutional layer and padding used to make the operation "causal"
kernel_size = 3
conv = torch.nn.Conv1d(in_channels=1, out_channels=1, kernel_size=kernel_size).to(
    device
)
pad = torch.nn.ZeroPad1d((kernel_size - 1, 0)).to(device)

# a length-12 sequence with 4 NaN elements at it's end
incomplete_sequence = torch.randn(1, 1, 12).to(device)
incomplete_sequence[:, :, -4:] = torch.nan
print(f"Test sequence has {incomplete_sequence.isnan().sum()} input NaNs.\n")

# test the operation for three different batch sizes
for batch_size in [4, 16, 32]:

    # a new batch with random data
    x = torch.randn(batch_size, 1, 12).to(device)

    # insert the incomplete sequence in position 3 (doesn't matter which position)
    x[3] = incomplete_sequence

    # apply padding and convolution that should NOT produce additional NaNs
    with torch.no_grad():
        y = conv(pad(x))

    print(
        f"Batch Size: {batch_size:<3} -> Output NaNs for the sequence: {y[3].isnan().sum()}"
    )
```

If I run this on `cpu`, I get the following output showing that extra NaNs appear only for the larger batches:
```
Test sequence has 4 input NaNs.

Batch Size: 4   -> Output NaNs for the sequence: 4
Batch Size: 16  -> Output NaNs for the sequence: 6
Batch Size: 32  -> Output NaNs for the sequence: 6
```

But if I run it on `mps`, the extra NaNs do not appear:
```
Test sequence has 4 input NaNs.

Batch Size: 4   -> Output NaNs for the sequence: 4
Batch Size: 16  -> Output NaNs for the sequence: 4
Batch Size: 32  -> Output NaNs for the sequence: 4
```

### Versions

PyTorch version: 2.6.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.5 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.0.13.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.7 (main, Oct 16 2024, 07:12:08) [Clang 18.1.8 ] (64-bit runtime)
Python platform: macOS-15.5-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Pro

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect

**Comment by ZailiWang** [comment]
Hi, I tested with 20250629 nightly build on a Xeon CPU, and the issue is not reproduced, looks like it's been fixed recently.
Would you verify with latest build again?

**Comment by markmbaum** [comment]
Can confirm that when `torch.__version__` is `2.9.0.dev20250630` the problem persists (on my laptop):
```
Test sequence has 4 input NaNs.

Batch Size: 4   -> Output NaNs for the sequence: 4
Batch Size: 16  -> Output NaNs for the sequence: 6
Batch Size: 32  -> Output NaNs for the sequence: 6
```

**Comment by hvaara** [comment]
I can also reproduce on an Apple CPU. `y[3]` in the case of batch size 32 with

Xeon CPU:
```
tensor([[ 0.2641,  0.1857,  0.1816,  0.5896,  0.3493,  0.5669, -0.1643,  0.7094,
             nan,     nan,     nan,     nan]])
```

CUDA:
```
tensor([[ 0.2641,  0.1857,  0.1816,  0.5896,  0.3493,  0.5669, -0.1643,  0.7094,
             nan,     nan,     nan,     nan]], device='cuda:0')
```

Apple M3 Max 128GB CPU:
```
tensor([[0.2641, 0.1857, 0.1816, 0.5896, 0.3493, 0.5669,    nan,    nan,    nan,
            nan,    nan,    nan]])
```

MPS:
```
tensor([[ 0.2641,  0.1857,  0.1816,  0.5896,  0.3493,  0.5669, -0.1643,  0.7094,
             nan,     nan,     nan,     nan]], device='mps:0')
```

Perhaps someone with a non-Apple ARM CPU can try to repro to determine if its ARM specific or not?

@pytorchbot label "needs reproduction" "module: arm"

**Comment by hvaara** [comment]
I was able to test with an a1.medium instance on AWS, and could not repro bad:

```
Batch Size: 4   -> Output NaNs for the sequence: 4
Batch Size: 16  -> Output NaNs for the sequence: 4
Batch Size: 32  -> Output NaNs for the sequence: 4
```

for `y[3]`:

```
tensor([[ 0.2641,  0.1857,  0.1816,  0.5896,  0.3493,  0.5669, -0.1643,  0.7094,
             nan,     nan,     nan,     nan]])
```

This does not seem related to ARM in general. Possibly Apple-specific?

**Comment by fadara01** [comment]
Edited: (I originally missed the point)

I tested this on my AArch64 machine (c8g) and I can't reproduce - i.e. the extra nans do not appear:
```
Test sequence has 4 input NaNs.

Batch Size: 4   -> Output NaNs for the sequence: 4
Batch Size: 16  -> Output NaNs for the sequence: 4
Batch Size: 32  -> Output NaNs for the sequence: 4
```

**Comment by hvaara** [comment]
Great, thanks for confirming!

I've been able to repro bad since at least PyTorch 2.3.0 with macOS 15.5. What I think is weird is that it's WAI on MPS, but broken on CPU. I'll add a test to see if this is reproducible across earlier macOS versions.

A couple more data points: I can't repro bad with a batch size below 16, but broken at and above. Not been able to repro the issue with dtypes float16, float64, bfloat16, complex32, or complex128, but I can repro bad with float32 and complex64.

Not sure if it's more correct to remove the arm label or not since this issue seems macOS specific and not directly related to ARM.

@pytorchbot label "module: correctness (silent)" "module: macos" "module: NaNs and Infs"

**Comment by markmbaum** [comment]
I see the same thing fwiw (again on macOS): extra NaNs only appear at batch size >15 and with `float32` and `complex64`.

**Comment by hvaara** [comment]
This is related to NNPACK. The reason the error does not trigger for batch sizes < 16 is the `at::symint::size<T>(input, 0) >= 16` check in

https://github.com/pytorch/pytorch/blob/19ae5afdaa3a59135e6adbaea197b7a2456f1865/aten/src/ATen/native/Convolution.cpp#L536-L549

Not sure exactly where the error is yet, but getting closer.

@pytorchbot label "module: nnpack"

**Comment by hvaara** [comment]
Building on linux arm from source I can repro bad.

`print(torch.__config__.show())` shows that `USE_MKLDNN` is `on` in the release version, but `off` when built locally.

on nightly:

```
PyTorch built with:
  - GCC 13.3
  - C++ Version: 201703
  - Intel(R) MKL-DNN v3.7.1 (Git Hash 8d263e693366ef8db40acc569cc7d8edf644556d)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: DEFAULT
  - Build settings: BLAS_INFO=open, BUILD_TYPE=Release, COMMIT_SHA=1c10a682eeca50952df72eed889efc6fcaf63526, CXX_COMPILER=/opt/rh/gcc-toolset-13/root/usr/bin/c++, CXX_FLAGS=-ffunction-sections -fdata-sections -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOCUPTI -DLIBKINETO_NOROCTRACER -DLIBKINETO_NOXPUPTI=ON -DUSE_PYTORCH_QNNPACK -DAT_BUILD_ARM_VEC256_WITH_SLEEF -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -O2 -fPIC -DC10_NODEPRECATED -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Werror=range-loop-construct -Werror=bool-operation -Wnarrowing -Wno-missing-field-initializers -Wno-unknown-pragmas -Wno-unused-parameter -Wno-strict-overflow -Wno-strict-aliasing -Wno-stringop-overflow -Wsuggest-override -Wno-psabi -Wno-error=old-style-cast -faligned-new -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-dangling-reference -Wno-error=dangling-reference -Wno-stringop-overflow, LAPACK_INFO=open, TORCH_VERSION=2.9.0, USE_CUDA=OFF, USE_CUDNN=OFF, USE_CUSPARSELT=OFF, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_GLOO=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=OFF, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, USE_ROCM_KERNEL_ASSERT=OFF, USE_XCCL=OFF, USE_XPU=OFF,
```

installed from source:

```
PyTorch built with:
  - GCC 13.3
  - C++ Version: 201703
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: DEFAULT
  - Build settings: BUILD_TYPE=Release, COMMIT_SHA=c52a002a220008f6cb13611d1cbc63f102e4ec40, CXX_COMPILER=/usr/bin/c++, CXX_FLAGS= -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOCUPTI -DLIBKINETO_NOROCTRACER -DLIBKINETO_NOXPUPTI=ON -DUSE_PYTORCH_QNNPACK -DAT_BUILD_ARM_VEC256_WITH_SLEEF -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -O2 -fPIC -DC10_NODEPRECATED -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Werror=range-loop-construct -Werror=bool-operation -Wnarrowing -Wno-missing-field-initializers -Wno-unknown-pragmas -Wno-unused-parameter -Wno-strict-overflow -Wno-strict-aliasing -Wno-stringop-overflow -Wsuggest-override -Wno-psabi -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-dangling-reference -Wno-error=dangling-reference -Wno-stringop-overflow, TORCH_VERSION=2.8.0, USE_CUDA=OFF, USE_CUDNN=OFF, USE_CUSPARSELT=OFF, USE_EIGEN_FOR_BLAS=ON, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_GLOO=ON, USE_MKL=OFF, USE_MKLDNN=OFF, USE_MPI=OFF, USE_NCCL=OFF, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, USE_ROCM_KERNEL_ASSERT=OFF, USE_XCCL=OFF, USE_XPU=OFF,
```

Installing from source with `USE_MKLDNN=1 python setup.py develop`, I can no longer repro bad.

I believe `USE_MKLDNN` is set for release versions of PyTorch on ARM for linux in https://github.com/pytorch/pytorch/blob/86eaf452c330bd871262b7590ff8bd1bf432e2d7/.ci/pytorch/build.sh#L96-L100

This however does not apply to the macOS release of PyTorch because the build environment is `macos-arm64-binary-wheel`, which does not contain `aarch64`.

https://github.com/pytorch/pytorch/blob/f7fd47f6e7456486e46af5076bd57e50c88aa140/.github/workflows/generated-macos-arm64-binary-wheel-nightly.yml#L24

`USE_MKLDNN` defaults to `off` when building from source on ARM (ie. not amd64/x86_64).

https://github.com/pytorch/pytorch/blob/f7fd47f6e7456486e46af5076bd57e50c88aa140/CMakeLists.txt#L313-L317

This explains the behavior we're seeing, but I'm unsure if this is intended or not.

@malfet Do you know why macOS is built without MKL? Should macOS releases be built with MKL?

**Comment by fadara01** [comment]
>I believe USE_MKLDNN is set for release versions of PyTorch on ARM for linux in

Yeah this is correct. This builds pytorch with oneDNN and Arm Compute Library for AArch64
